src/Storage/Notmuch.hs

```diff
@@ -34,15 +34,34 @@ getMessages :: T.Text
             -> NotmuchSettings FilePath
             -> IO (Either String (Vec.Vector NotmuchMail))
 getMessages s settings =
-  first show <$>
-  bracket (runExceptT (databaseOpenReadOnly (view nmDatabase settings)))
-          (runExceptT . traverse_ databaseDestroy)
-          (runExceptT . (either throwError pure >=> go))
+  first show
+  <$> withRODatabase (view nmDatabase settings) go
   where go db = do
               msgs <- query db (FreeForm $ T.unpack s) >>= messages
               mails <- liftIO $ mapM messageToMail msgs
               pure $ Vec.fromList mails
 
+mailFilepath :: NotmuchMail -> FilePath -> IO (Either String FilePath)
+mailFilepath m dbpath =
+    withRODatabase dbpath go >>=
+    either
+        (pure . Left . show)
+        (pure .
+         maybe
+             (Left "No message found. Is your notmuch correctly setup?")
+             Right)
+  where
+    go db = findMessage db (view mailId m) >>= traverse messageFilename
```

Comment:
can tidy this one up a bit:
```haskell
mailFilepath :: NotmuchMail -> FilePath -> IO (Either String FilePath)              
mailFilepath m dbpath =                                                             
    join . bimap show (maybe (Left notFoundMsg) Right) <$> withRODatabase dbpath go 
  where                                                                             
    notFoundMsg = "No message found. Is your notmuch correctly setup?"              
    go db = findMessage db (view mailId m) >>= traverse messageFilename             
```

Again, the forthcoming refactoring should improve things a bit more.